sql/schemachanger: Preserving column family order for complex column type changes

spilchen opened the pull request:
When altering a column’s type in a way that requires a backfill, we drop the old column and add a new one. Previously, the new column was always added to the end of the column family. This change ensures the column family order is preserved.

In the DSC, the column family is updated when a new ColumnType element is added. I introduced a new field to this element to control the order, which requires specifying the column ID that the new column should follow.

Epic: CRDB-25314
Closes #133040
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133845)
<!-- Reviewable:end -->

fqazi reviewed (changes requested):
> Reviewed 1 of 14 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133845)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133845#-OAb__PA-OUEeVB8jX4K):*
Trying to understand the rationale for why we need this? From a SQL viewpoint nothing changes based on the order (maybe the DDL from SHOW CREATE only). I think a hazard is that rows could become unreadable if we aren't careful about how this is ordered


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133845)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133845#-OAb__PA-OUEeVB8jX4K):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Trying to understand the rationale for why we need this? From a SQL viewpoint nothing changes based on the order (maybe the DDL from SHOW CREATE only). I think a hazard is that rows could become unreadable if we aren't careful about how this is ordered
</blockquote></details>

The reasons were:
-   To ensure consistent output in commands like `SHOW CREATE` after the alteration.
-   The legacy schema changer also preserved the order.

I admit these aren’t strong reasons, so I’m okay with deferring this work.


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
> Reviewed 13 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133845)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133845#-OAb__PA-OUEeVB8jX4K):*
<details><summary><i>Previously, spilchen wrote…</i></summary><blockquote>

The reasons were:
-   To ensure consistent output in commands like `SHOW CREATE` after the alteration.
-   The legacy schema changer also preserved the order.

I admit these aren’t strong reasons, so I’m okay with deferring this work.
</blockquote></details>

Just as sanity lets confirm this with a DML injection test to confirm that everything stays readable. The changes themselves look fine, but it would just add piece of mind.


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/schemachanger: Preserving column family order for complex column type changes

When altering a column’s type in a way that requires a backfill, we drop
the old column and add a new one. Previously, the new column was always
added to the end of the column family. This change ensures the column
family order is preserved.

In the DSC, the column family is updated when a new ColumnType element
is added. I introduced a new field to this element to control the order,
which requires specifying the column ID that the new column should
follow.

Epic: CRDB-25314
Closes #133040
Release note: none

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133845)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi and @spilchen)

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133845#-OAb__PA-OUEeVB8jX4K):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Just as sanity lets confirm this with a DML injection test to confirm that everything stays readable. The changes themselves look fine, but it would just add piece of mind.
</blockquote></details>

I added a DML injection test. But there is a known problem where the column we are altering can become invisible for a couple of stages. I have issue #133996 opened to investigate that.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTR @fqazi 

bors r+
